### DISCUSSION TITLE
How to verify the execution proof of an extrinsic

### DISCUSSION THREAD

**Discussion by liuchengxu** [paritytech/substrate#10987] [Q&A] [answered]
Generally speaking, in order to verify a state transition, we need the initial state, extrinsic(data), and a runtime(state transition function), apply the state transition on our own, and then check if the locally produced `post_state_root` is the same as the one received from the outside.

Currently `execution_proof_check` only returns the execution result which is not enough for supporting the extrinsic execution proof (https://github.com/paritytech/substrate/discussions/10783).

https://github.com/paritytech/substrate/blob/8ce691dcc163c4ca3fe90a33f87c8821d1a73412/primitives/state-machine/src/lib.rs#L630-L640

Please correct me if I'm wrong, the block execution validation in polkadot is done by re-executing the parachain block and producing a new block header which will be used to compare with the one received from the collator,  hence returning the execution result is enough as the header includes the state root already. However, to verify the execution proof of a certain extrinsic, `post_state_root` is necessary since we can't compare the two executions with only the execution result.

Is there any way to also return the `post_state_root` when verifying an execution proof?

**Comment by bkchr** [comment]
> Please correct me if I'm wrong, the block execution validation in polkadot is done by re-executing the parachain block and producing a new block header which will be used to compare with the one received from the collator, hence returning the execution result is enough as the header includes the state root already.

Parachains don't use this code path here.


> However, to verify the execution proof of a certain extrinsic, `post_state_root` is necessary since we can't compare the two executions with only the execution result.

You check an execution proof, by re-doing the same execution with the given proof. If that works, it means that the given proof is valid. Not sure why you want to compare the state roots? How should you get a different state root if you execute the same code?

**Reply by liuchengxu** [reply]
Actually _the proof is valid_ only indicates the method and the call data can be applied on top of the state constructed from the storage proof, that does not mean they follow the identical execution path. _How should you get a different state root if you execute the same code?_ Yeah, that's the problem, the code can be different if an executor is malicious :D.

**Comment by liuchengxu** [comment, marked as answer]
Thanks to @cheme suggested in https://github.com/paritytech/substrate/pull/10922#issuecomment-1066752530, take `apply_extrinsic` as an example, one solution here is to an identical API of `apply_extrinsic` but call the storage root at the end and use it as the return value(might be useful to have a look at https://github.com/paritytech/substrate/issues/11046 to decode the return value properly). And then we can interpret the return value as `post_state_root` and compare it with the one received from the network.
